### DISCUSSION TITLE
Gambit REPL in Emacs

### DISCUSSION THREAD

**Discussion by jcubic** [gambit/gambit#896] [Q&A] [answered]
In Emacs, you can run scheme interpreter is special interactive mode. In any scheme I was testing, when you execute the code, it just inserts the prompt and don't add a new line. I fixed this by injecting code `(newline)` after evaluating expression from Scheme buffer.

But I've noticed that Gambit works differently in Emacs, the code you execute is visible in the interpreter.

This is Guile:

![scheme-mode-01](https://github.com/gambit/gambit/assets/280241/c3f53e12-31df-4813-8165-1e101af78e47)

This is Gabmit:

![scheme-mode-02](https://github.com/gambit/gambit/assets/280241/d5cd7c52-69e0-461d-a35a-1971f82160fc)

What Gambit REPL is doing to achieve this effect? Notice the visible `(newline)` at the end that is sent to the REPL.

I know that I can read the code, but it is probably faster to ask if someone already know the answer.

**Comment by feeley** [comment]
I assume you are using `gambit.el`. This part is probably doing this:

```
;; Redefine the function scheme-send-region from `cmuscheme' so
;; that we can keep track of all text sent to Gambit's stdin.

(defun scheme-send-region (start end)
  "Send the current region to the inferior Scheme process."
  (interactive "r")
  (scheme-send-string (buffer-substring start end)))

(defun scheme-send-string (str)
  "Send a string to the inferior Scheme process."
  (let* ((clean-str (gambit-string-terminate-with-newline str))
         (proc (scheme-proc))
         (pmark (process-mark proc))
         (buffer (get-buffer scheme-buffer))
         (old-buffer (current-buffer)))
    (set-buffer buffer)
    (goto-char pmark)
    (set-marker comint-last-input-start (point))
    (insert clean-str)
    (set-marker pmark (point))
    (gambit-input-sender proc clean-str)
    (set-buffer old-buffer)))

(defun gambit-input-sender (proc str)
  (let ((clean-str (gambit-string-terminate-with-newline str)))
    (gambit-register-input clean-str)
    (gambit-make-read-only (current-buffer) (point-max))
    (gambit-unhighlight)
    (comint-send-string proc clean-str)))
```

**Reply by jcubic** [reply]
No, I don't use any specific for Gambit in Emacs, I use generic Scheme mode. So Gambit REPL needs to be doing something specific to allow this.

This is my setup:

```lisp
(require 'cmuscheme)
(autoload 'scheme-mode "cmuscheme" "Major mode for Scheme." t)
(autoload 'run-scheme "cmuscheme" "Switch to interactive Scheme buffer." t)
(add-hook 'scheme-mode-hook 'turn-on-font-lock)

(defun gambit ()
  "call run-scheme with gambit interpreter."
  (interactive)
  (run-scheme "/usr/bin/env gsi"))

(defun scheme-send-last-sexp-fix ()
  "Send the previous sexp to the inferior Scheme process
   and add newline at the end"
  (interactive)
  (scheme-send-last-sexp)
  (comint-send-string (scheme-proc) "(newline)\n")
  )

(define-key scheme-mode-map "\C-x\C-e" #'scheme-send-last-sexp-fix)
```

**Comment by feeley** [comment, marked as answer]
Perhaps guile is doing something to prevent the echo. Is guile sensitive to the `INSIDE_EMACS` environment variable? Gambit is sensitive to this variable and disables the echo in that case, and I think emacs detects the absence of echo and then copies the input in that case to simulate the echo.

**Reply by jcubic** [reply]
Thanks, the same happen in few other Scheme interpreters I tested.

**Reply by jcubic** [reply]
I just found that Gambit don't do anything like this. It just prints what it echoes in Emacs. So if you type something in REPL you get the same expression twice.

**Reply by lassik** [reply]
I get twice printing in Emacs `M-x shell` too. The cure is `gsi -:tE` (disable terminal line editing). (Try `gsi -:help` for usage.)

**Reply by lassik** [reply]
`(get-environment-variable "TERM") ; => "dumb"` in `M-x shell`. Perhaps the REPL could auto-detect this.

**Comment by maconcg** [comment]
@jcubic I'm just now seeing this as I hadn't looked at the "Discussions" page.  I recently submitted some patches (#922) that I believe will affect the issue you've described here.  If you're able, would you mind trying it out and verifying whether things work as they "should" for you?

**Reply by jcubic** [reply]
I will wait when you finish the fix, and it will land on the master branch.
